### ISSUE TITLE
A bug in Function IsHalfSubnormal

### ISSUE THREAD

**Issue by ian37** [KhronosGroup/OpenCL-CTS#534]
https://github.com/KhronosGroup/OpenCL-CTS/blob/81cffe91326fef77d819de51ac37bde5795723cd/test_conformance/half/cl_utils.h#L117

If reference result is sub normal, the CTS case will test if the output is flushed to zero with the code below:
`if( IsHalfSubnormal(u2[j]) && ( (u1[j] == 0) || (u1[j] == 0x8000) ) )
    continue;`
u2 is initialized as 0x0400U when it is a denormalized number. The related code in function float2half_rte is as below:
`// half denormal`
`if( x < MAKE_HEX_FLOAT(0x1.0p-14f, 0x1L, -14) )
{
    x *= MAKE_HEX_FLOAT(0x1.0p24f, 0x1L, 24);
    return (cl_ushort)((int) x | sign);
}`
However, the function IsHalfSubnormal uses 0x03ffU instead of 0x0400U. 
`static inline int IsHalfSubnormal( uint16_t x )
{
    return ((x&0x7fffU)-1U) < 0x03ffU; 
}`
In my option, the 0x03ffU should be 0x0400U.

**Comment by jlewis-austin** [comment]
What does the MAKE_HEX_FLOAT code have to do with the issue? It looks like the code shown comes from float2half_rtz, not float2half_rte. 
The IsHalfSubnormal routine threshold looks correct to me, set to 0x03ffU to account for the subtraction by 1U. An input of 0x0400 (smallest normal) will become 0x03ff after subtraction, which is not strictly less than 0x03ffU, so it returns 0. Anything smaller nonzero value is a subnormal. Zero is excluded because the subtraction rolls over to 0xffffU. Can you provide an example of an input that returns an incorrect result?

**Comment by ian37** [comment]
First, sorry for my 2 mistakes.
1. I pasted the wrong code. I should have pasted the code from float2half_rte.
2. It is a problem in the cl12_trunk. But I added the code link that links to the code in the master.

Please let me re-describe the problem.

Let's see float2half_rte(, which is in test_conformance/half/Test_vStoreHalf.c).
https://github.com/KhronosGroup/OpenCL-CTS/blob/2c0abd3f75d34be97682cd410a1171b87efe964c/test_conformance/half/Test_vStoreHalf.c#L61
cl_ushort float2half_rte( float f )
{
......
    // half denormal
    if( x < MAKE_HEX_FLOAT(0x1.0p-14f, 0x1L, -14) )
    {
        u.f = x * MAKE_HEX_FLOAT(0x1.0p-125f, 0x1L, -125);
        return sign | u.u;
    }
......
}
When the input value is 0x1.ffcp-15, which is less than 0x1.0p-14f, float2half_rte() will go into the half denormal branch and return 0x0400 after rounding.
Then, the u2 in the code below(, test_conformance/half/Test_vStoreHalf.c: line 958,) is 0x0400, which is the smallest normal.
if( IsHalfSubnormal(u2[j]) && ( (u1[j] == 0) || (u1[j] == 0x8000) ) )  continue;
https://github.com/KhronosGroup/OpenCL-CTS/blob/2c0abd3f75d34be97682cd410a1171b87efe964c/test_conformance/half/Test_vStoreHalf.c#L958
And if the hardware does not support denorm, the hardware can directly flush the input value 0x1.ffcp-15 to 0, which is described in Specification:
"Denormalized numbers for the half data type which may be generated when converting a float to a half using variants of the vstore_half function or when converting from a half to a float using variants of the vload_half function can be flushed to zero."
So, when input value is 0x1.ffcp-15, u1 is 0.
In summary, when input is 0x1.ffcp-15 and the hardware does not support denorm, u1 is 0 and u2 is 0x0400.
Then it is impossible to go into the branch "continue".

There are 2 options to deal with this problem:
Option 1: The function IsHalfSubnormal should use 0x0400U instead of 0x03ffU:
https://github.com/KhronosGroup/OpenCL-CTS/blob/2c0abd3f75d34be97682cd410a1171b87efe964c/test_conformance/half/cl_utils.h#L137
Option 2: u2 is the smallest normal after rounding. It should not be the input parameter of IsHalfSubnormal. The input parameter of IsHalfSubnormal should be the number before rounding.

**Comment by jlewis-austin** [comment]
OK got it.  I disagree with Option 1 because IsHalfSubnormal is correct as written for the half input it's given--if we want to account for rounding from floats, that needs to be a new function that takes a float input.
I think Option 2 looks like the right fix (using the original float input with a new function to detect denormals before rounding). This problem is in both cl12_trunk and master. cl12_trunk also is missing the gIsEmbedded check, so it incorrectly allows flushing on any half conversion.

**Comment by rjodinchr** [comment]
Obsolete: As confirmed by CTS maintainers, `IsHalfSubnormal` is mathematically correct per IEEE 754-2008 for `cl_half` inputs (`((x & 0x7fffU) - 1U) < 0x03ffU`), and the issue author acknowledged misinterpreting code from an older branch.
